### ISSUE TITLE
Add Feature: Attach Application EPGs to AAEP

### ISSUE THREAD

**Issue by aj-cruz** [CiscoDevNet/terraform-provider-aci#67]
I've been looking through the provider documentation but I don't see a way to add application EPGs directly to an AAEP (class infraRsFuncToEpg I think).

**Comment by nkatarmal-crest** [comment]
@aj-cruz I think you are looking for this. https://www.terraform.io/docs/providers/aci/r/infraprovacc.html#relation_infra_rs_func_to_epg.

**Comment by aj-cruz** [comment]
@nkatarmal-crest I don't know if that's the right one, the name is strange but the class looks right.
However, if it is supposed to be used for the EPG-to-AAEP relation, it is missing a couple important attributes:
encap
mode

I went ahead and tried it, here's what the plan looks like:
```
Terraform will perform the following actions:

  # module.tier3.aci_vlan_encapsulationfor_vxlan_traffic.aaep["relation1"] will be created
  + resource "aci_vlan_encapsulationfor_vxlan_traffic" "aaep" {
      + annotation                          = (known after apply)
      + attachable_access_entity_profile_dn = "uni/infra/attentp-Physical-AAEP"
      + description                         = (known after apply)
      + id                                  = (known after apply)
      + name_alias                          = (known after apply)
      + relation_infra_rs_func_to_epg       = [
          + "uni/tn-TENANT1/ap-Production-Network/epg-VLAN5-EPG",
        ]
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

Here are the results:
```
module.tier3.aci_vlan_encapsulationfor_vxlan_traffic.aaep["relation1"]: Creating...
module.tier3.aci_vlan_encapsulationfor_vxlan_traffic.aaep["relation1"]: Creation complete after 0s [id=uni/infra/attentp-Physical-AAEP/provacc]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```

But no relation is made in ACI. Here's the trace:
```
2020-06-24T07:50:43.333-0500 [DEBUG] plugin.terraform-provider-aci: 2020/06/24 07:50:43 [DEBUG] Exit from errors {"imdata":[],"totalCount":"0"}
2020-06-24T07:50:43.333-0500 [DEBUG] plugin.terraform-provider-aci: 2020/06/24 07:50:43 HTTP request POST /api/node/mo.json &{POST https://192.168.253.17/api/node/mo.json HTTP/1.1 1 1 map[] {{"infraRsFuncToEpg":{"attributes":{"dn":"uni/infra/attentp-Physical-AAEP/provacc/provacc/rsfuncToEpg-[uni/tn-TENANT1/ap-Production-Network/epg-VLAN5-EPG]"}}}} 0x765d90 157 [] false 192.168.253.17 map[] map[] <nil> map[]   <nil> <nil> <nil> 0xc00003a0a8}
2020-06-24T07:50:43.333-0500 [DEBUG] plugin.terraform-provider-aci: 2020/06/24 07:50:43 [DEBUG] Begin Injection
2020-06-24T07:50:43.333-0500 [DEBUG] plugin.terraform-provider-aci: 2020/06/24 07:50:43 HTTP request after injection POST /api/node/mo.json &{POST https://192.168.253.17/api/node/mo.json HTTP/1.1 1 1 map[Cookie:[APIC-Cookie=owUAAAAAAAAAAAAAAAAAACgeG5+R8TwtOanVHaD0lq66CEBmOBsq45QbkyBTe/K5dxDJoedvg9aeC5ql5Xm8WRiaueE4YFeiM4qX1eLXTy6sH9ajeNOHOluhbqWA+kNJuKQqgD78Q2Odeg4793/XOqAWCO6DCoaR6bAzrVQ9vhx+wBT9Y6rywpXNvRTqIjEC259/8yb0iqfL4521eHihDg==]] {{"infraRsFuncToEpg":{"attributes":{"dn":"uni/infra/attentp-Physical-AAEP/provacc/provacc/rsfuncToEpg-[uni/tn-TENANT1/ap-Production-Network/epg-VLAN5-EPG]"}}}} 0x765d90 157 [] false 192.168.253.17 map[] map[] <nil> map[]   <nil> <nil> <nil> 0xc00003a0a8}
2020-06-24T07:50:43.333-0500 [DEBUG] plugin.terraform-provider-aci: 2020/06/24 07:50:43 [DEBUG] Begining DO method https://192.168.253.17/api/node/mo.json
2020-06-24T07:50:43.348-0500 [DEBUG] plugin.terraform-provider-aci: 2020/06/24 07:50:43 
2020-06-24T07:50:43.348-0500 [DEBUG] plugin.terraform-provider-aci: 
2020-06-24T07:50:43.348-0500 [DEBUG] plugin.terraform-provider-aci: 
2020-06-24T07:50:43.348-0500 [DEBUG] plugin.terraform-provider-aci:  HTTP request: {}
2020-06-24T07:50:43.348-0500 [DEBUG] plugin.terraform-provider-aci: 2020/06/24 07:50:43 
2020-06-24T07:50:43.348-0500 [DEBUG] plugin.terraform-provider-aci: HTTP Request: POST https://192.168.253.17/api/node/mo.json
2020-06-24T07:50:43.348-0500 [DEBUG] plugin.terraform-provider-aci: 2020/06/24 07:50:43 nHTTP Response: 400 400 Bad Request &{400 Bad Request 400 HTTP/1.1 1 1 map[Access-Control-Allow-Credentials:[false] Access-Control-Allow-Headers:[Origin, X-Requested-With, Content-Type, Accept, devcookie, APIC-challenge, Request-Tag] Access-Control-Allow-Methods:[POST,GET,OPTIONS,DELETE] Access-Control-Allow-Origin:[http://127.0.0.1:8000] Connection:[keep-alive] Content-Length:[270] Content-Type:[application/json] Date:[Thu, 11 Jun 2020 13:21:10 GMT] Server:[Cisco APIC]] 0xc0001af340 270 [] false false map[] 0xc0001fc800 0xc00077cfd0}
2020-06-24T07:50:43.348-0500 [DEBUG] plugin.terraform-provider-aci: 2020/06/24 07:50:43 
2020-06-24T07:50:43.348-0500 [DEBUG] plugin.terraform-provider-aci:  HTTP response unique string POST https://192.168.253.17/api/node/mo.json {"totalCount":"1","imdata":[{"error":{"attributes":{"code":"120","text":"unknown property value uni\/infra\/attentp-Physical-AAEP\/provacc\/provacc\/rsfuncToEpg-[uni\/tn-TENANT1\/ap-Production-Network\/epg-VLAN5-EPG], name dn, class infraRsFuncToEpg [(Dn0)] Dn0=, "}}}]}
2020-06-24T07:50:43.348-0500 [DEBUG] plugin.terraform-provider-aci: 2020/06/24 07:50:43 [DEBUG] Exit from do method
```

And here's what a good post looks like, adding the relation manually, from the API Inspector:
```
method: POST
url: https://apic/api/node/mo/uni/infra/attentp-Physical-AAEP/gen-default.json
payload{"infraGeneric":{"attributes":{"dn":"uni/infra/attentp-Physical-AAEP/gen-default","name":"default","status":"created,modified"},"children":[{"infraRsFuncToEpg":{"attributes":{"tDn":"uni/tn-TENANT1/ap-Production-Network/epg-VLAN5-EPG","status":"created,modified","encap":"vlan-5"},"children":[]}}]}}
response: {"totalCount":"0","imdata":[]}
```

**Comment by nkatarmal-crest** [comment]
@aj-cruz I will add a new relation resource, so that we can manage other parameters as well.

**Comment by poomnupong** [comment]
Has been looking for this too.
Here's workaround I'm using with aci_rest. Interested in seeing if there's a better way.
aci_rest doesn't seem to track state so I have wrapper script that taint this resource before deploy.
```
# must taint with before apply:
# > terraform taint aci_rest.bind_vl697epg_to_aep
resource "aci_rest" "bind_vl697epg_to_aep" {
  path       = "/api/node/mo/${aci_attachable_access_entity_profile.irvsdc2-fi1_aep.id}/gen-default.json"
  class_name = "infraRsFuncToEpg"
  content = {
    "tDn" : aci_application_epg.vl697-10_95_19_96_27_epg.id,
    "status" : "created,modified"
    "encap" : "vlan-697"
  }
  depends_on = [aci_application_epg.vl697-10_95_19_96_27_epg, aci_attachable_access_entity_profile.irvsdc2-fi1_aep]
}
```

**Comment by Bullenj01** [comment]
I had created my own module for this because I didn't find this resource till I happened across this Issue.  I would say a rename of this function may be needed because when people are looking for it they are going to search for something that has epg and aep.  Nobody is going to know that aci_epgs_using_function maps an EPG to an AEP.  First block is my original module that I've been using in prod and works well, it obfuscates the infraGeneric class.  the second code block is an example using the canon resources that seems to work just fine too

```hcl
variable "aep" {
  description = "Name of existing aep you want to add an epg to"
  type        = string
}
variable "epg" {
  description = "Name of existing epg you want to add to aep above"
  type        = string
}
variable "vlan" {
  description = "Vlan tag to use for this encap"
  type        = string
}

resource "aci_rest" "epg_to_aep" {
  path       = "/api/node/mo/uni/infra/attentp-${var.aep}/gen-default.json"
  #class_name = "infraRsFuncToEpg"
  class_name = "infraGeneric"
  payload = <<EOF
{
  "infraGeneric": {
        "attributes": {
            "dn": "uni/infra/attentp-vmware/gen-default",
            "name": "default",
        },
        "children": [
            {
                "infraRsFuncToEpg": {
                    "attributes": {
                        "tDn": "${var.epg}",
                        "status": "created,modified",
                        "encap": "vlan-${var.vlan}"
                    },
                    "children": []
                }
            }
        ]
    }
}
  EOF
}
```

```hcl
variable "aep" {
  description = "Name of existing aep you want to add an epg to"
  type        = string
}
variable "epg" {
  description = "Name of existing epg you want to add to aep above"
  type        = string
}
variable "vlan" {
  description = "Vlan tag to use for this encap"
  type        = string
}
data "aci_attachable_access_entity_profile" "this" {
  name = var.aep
}
resource "aci_access_generic" "this" {
  attachable_access_entity_profile_dn = data.aci_attachable_access_entity_profile.this.id
  name = "default"
  annotation = "orchestrator:terraform"
}
resource "aci_epgs_using_function" "epg_to_aep" {
  access_generic_dn = aci_access_generic.this.id
  tdn = var.epg
  encap = "vlan-${var.vlan}"
  annotation = "orchestrator:terraform"
  #instr_imedcy =
  #mode =
  #primary_encap =
}
```

Example of a call

```terraform
module "vmware_mgmt-vmware" {
  source = "../../modules/epg_to_aep"
  aep = "vmware"
  epg = aci_application_epg.vmware_mgmt.id
  vlan = "6"
}
```

**Comment by nkatarmal-crest** [comment]
This issue has been fixed in the version 0.5.0 of the provider. Please try that and let me know if we can close this issue or not.
